File: Examenes-EDD-2/2020-Septiembre/Haskell/AVL.hs
 {------------------------------------------------------------------------------
 - Student's name: Paco Molina Cuenca
 -
 - Student's group: Doble Grado Matemáticas e Ingeniería Informática.
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin capacity = B capacity []

remainingCapacity :: Bin -> Capacity
remainingCapacity (B capacity _) = capacity

addObject :: Weight -> Bin -> Bin
addObject w (B capacity ws)
  | w > capacity  = error "Too heavy object for the given bin"
  | otherwise     = B (capacity - w) (w:ws) 

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty                    = 0
maxRemainingCapacity (Node _ _ capacity _ _)  = capacity

height :: AVL -> Int
height Empty            = 0
height (Node _ h _ _ _) = h

nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight bin height lt rt = Node bin height (maximum [remainingCapacity bin, maxRemainingCapacity lt, maxRemainingCapacity rt]) lt rt

node :: Bin -> AVL -> AVL -> AVL
node bin lt rt = nodeWithHeight bin (1 + max (height lt) (height rt)) lt rt

rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft c l (Node x _ _ r1 r2) = node x (node c l r1) r2  

addNewBin :: Bin -> AVL -> AVL
addNewBin bin Empty = Node bin 1 (remainingCapacity bin) Empty Empty
addNewBin bin (Node x h maxCapacity lt rt)
  | height rightChild > height lt + 1 = rotateLeft x lt rightChild
  | otherwise                         = node x lt rightChild
  where 
    rightChild = addNewBin bin rt

addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst capacity w Empty         = Node (B (capacity - w) [w]) 1 (capacity - w) Empty Empty
addFirst capacity w (Node bin height maxCapacity lt rt)
  | maxRemainingCapacity lt >= w  = node bin (addFirst capacity w lt) rt 
  | remainingCapacity bin >= w    = node (addObject w bin) lt rt
  | maxRemainingCapacity rt >= w  = node bin lt (addFirst capacity w rt)
  | otherwise                     = node bin lt (addNewBin (B (capacity - w) [w]) rt)

addAll:: Capacity -> [Weight] -> AVL
addAll capacity = foldr (addFirst capacity) Empty 

toList :: AVL -> [Bin]
toList Empty = []
toList (Node bin _ _ lt rt) = toList lt ++ [bin] ++ toList rt 

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
-}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking capacity = foldl insert SEmpty
  where 
    insert :: Sequence -> Weight -> Sequence 
    insert SEmpty w                   = SNode (B (capacity - w) [w]) SEmpty
    insert (SNode bin seq) w 
      | remainingCapacity bin >= w    = SNode (addObject w bin) seq
      | otherwise                     = SNode bin (insert seq w)

seqToList:: Sequence -> [Bin]
seqToList SEmpty          = []
seqToList (SNode bin seq) = bin : seqToList seq

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold weights capacity = foldr (addFirst capacity) Empty weights 

{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"

File: Examenes-EDD-2/2020-Septiembre/Java/AVL.java
/**
 * Student's name: Paco Molina Cuenca
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 */

import dataStructures.list.LinkedList;
import dataStructures.list.List;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        remainingCapacity = initialCapacity;
        weights = new LinkedList<>();
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        return remainingCapacity;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        if(weight > remainingCapacity)
            throw new IllegalArgumentException("Object too heavy for the given bin");
        
        remainingCapacity -= weight;
        weights.append(weight);
    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        return weights.iterator();
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
            height = 1 + Math.max(height(left), height(right));
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
            maxRemainingCapacity = Math.max(Math.max(bin.remainingCapacity(), maxRemainingCapacity(left)), maxRemainingCapacity(right));
        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
            Node newRoot = right;
            right = newRoot.left;
            newRoot.left = this;

            // Actualizar alturas y maxRemainingCapacity
            this.setHeight();
            this.setMaxRemainingCapacity();
            newRoot.setHeight();
            newRoot.setMaxRemainingCapacity();

            return newRoot;
        }

    }
    
    private static int maxRemainingCapacity(Node node) {
        return node == null ? 0 : node.maxRemainingCapacity;
    }

    private static int height(Node node) {
        return node == null ? 0 : node.height;
    }

    private Node root; // root of AVL tree

    public AVL() {
        root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        root = addNewBinRec(bin, root);
    }

    private Node addNewBinRec(Bin bin, Node root) {
        if(root == null){
            Node newNode = new Node();
            newNode.bin = bin;
            newNode.left = null;
            newNode.right = null;
            newNode.setHeight();
            newNode.setMaxRemainingCapacity();
            
            return newNode;

        } else{
            root.right = addNewBinRec(bin, root.right);

            if(height(root.right) > height(root.left) + 1)
                root = root.rotateLeft();
            
            root.setHeight();
            root.setMaxRemainingCapacity();

            return root;
        }
    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
        root = addFirstRec(initialCapacity, weight, root);
    }

    private Node addFirstRec(int initialCapacity, int weight, Node root) {
        if(root == null){
            Node newNode = new Node();
            newNode.bin = new Bin(initialCapacity);
            newNode.bin.addObject(weight);
            newNode.left = null;
            newNode.right = null;
            newNode.setHeight();
            newNode.setMaxRemainingCapacity();
            return newNode;
        
        } else{
            if(root.left != null && root.left.maxRemainingCapacity >= weight){
                root.left = addFirstRec(initialCapacity, weight, root.left);

            }else if(root.bin != null && root.bin.remainingCapacity() >= weight){
                root.bin.addObject(weight);

            }else if(root.right != null && root.right.maxRemainingCapacity >= weight){
                root.right = addFirstRec(initialCapacity, weight, root.right);

            }else{ 
                Bin bin = new Bin(initialCapacity);
                bin.addObject(weight);
                root.right = addNewBinRec(bin, root.right);
            }

            return root;
        }
    }

    public void addAll(int initialCapacity, int[] weights) {
        for(int w : weights)
            addFirst(initialCapacity, w);
    }

    public List<Bin> toList() {
        return toListRec(root);
    }

    private List<Bin> toListRec(Node root) {
        List<Bin> resultado = new LinkedList<>();

        if(root == null){
            // Nada
        } else if(root.left == null && root.right == null){
            resultado.append(root.bin);
        
        } else{
            for(Bin cubo : toListRec(root.left))
                resultado.append(cubo);

            resultado.append(root.bin);

            for(Bin cubo : toListRec(root.right))
                resultado.append(cubo);
        }

        return resultado;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        List<Bin> res = new LinkedList<>();

        for(Integer w : weights)
            linearBinPackingRec(initialCapacity, w, res);

        return res;
    }

    public static void linearBinPackingRec(int initialCapacity, Integer weight, List<Bin> res){
        boolean isSettedUp = false;
        int i = 0;

        while(i < res.size() && !isSettedUp){
            if(res.get(i).remainingCapacity() >= weight){
                isSettedUp = true;
                res.get(i).addObject(weight);
            }
            i++;
        }

        if(!isSettedUp){
            Bin newBin = new Bin(initialCapacity);
            newBin.addObject(weight);
            res.append(newBin);
        }
    }

	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        List<Integer> aux = new LinkedList<>();

        for(Bin bin : bins){
            for(Integer weight : bin.objects())
                aux.append(weight);
        }

        return aux.iterator();
	}
}
